### PR TITLE
better reporting of GLSL compile messages

### PR DESCRIPTION
**Pull request by doug-moen**
If there are warning messages, but no error messages, we now report
the warning messages.

We report the pathname of the GLSL source file that contains the warnings
or errors (in case more than one shader is provided).

Line numbers reported in warning or error messages are now more accurate.
After the automatically generated prolog, a '#line 1' directive
is now inserted.

Fixed warnings in "mandelbrot.frag".

### CODE REVIEW

**Comment by patriciogonzalezvivo** [comment]
this is awesome!! keep the PR comming : ) Thanks again

File: src/gl/shader.cpp [hunk 1]
```diff
@@ -44,23 +44,19 @@ bool find_id(const std::string& program, const char* id) {
     return std::strstr(program.c_str(), id) != 0;
 }
 
-bool Shader::load(const std::string& _fragmentSrc, const std::string& _vertexSrc, bool verbose) {
+bool Shader::load(const std::string* _fragmentPath, const std::string& _fragmentSrc, const std::string* _vertexPath, const std::string& _vertexSrc, bool verbose) {
     std::chrono::time_point<std::chrono::steady_clock> start_time, end_time;
     start_time = std::chrono::steady_clock::now();
 
-    m_vertexShader = compileShader(_vertexSrc, GL_VERTEX_SHADER);
+    m_vertexShader = compileShader(_vertexPath, _vertexSrc, GL_VERTEX_SHADER);
```
Comment (patriciogonzalezvivo) [inline src/gl/shader.cpp]: have you test this with `#include "something.glsl"` ??

File: src/main.cpp [hunk 2]
```diff
@@ -429,8 +431,9 @@ void setup() {
     //  Build shader;
     //
     if (iFrag != -1) {
+        fragPath = &files[iFrag].path;
```
Comment (patriciogonzalezvivo) [inline src/main.cpp]: I have some project that use the `#include "something.glsl"` recursively. That is shaders than call `snip1.glsl` that call `snip1_1.glsl` etc... have you test it in those cases? I'm afraid of braking something here

File: src/gl/shader.cpp (hunk 1 above)
Comment (doug-moen) [inline src/gl/shader.cpp]: If `#include` is used, then (currently) line numbers will be incorrect. And I have no way to report the names of included files correctly in warning/error messages. That is because the GLSL standard doesn't permit you to put a pathname into a `#line` directive. So it's hard to fix this problem.

The _vertexPath and _shaderPath arguments of compileShader are the pathnames of the top level *.vert and *.frag files, which are globally unique. These pathnames are used only when printing messages.

I haven't tested `#include` yet. I will.

File: src/main.cpp (hunk 2 above)
Comment (doug-moen) [inline src/main.cpp]: You are only allowed to specify a maximum of 1 *.frag file on the command line. `fragPath` is the name of this *.frag file, or it is nullptr if not specified. The `files` array can contain a maximum of 1 fragment file, whose index is the global variable `iFrag` (-1 if there is no frag file). There is no code for watching if an included file has changed, and included files are not listed in the `files` array.

The code you highlighted is not executed in the case of `#include` processing. This is localized inside of `loadFromPath()` in `utils.h`.

File: src/gl/shader.cpp (hunk 1 above)
Comment (doug-moen) [inline src/gl/shader.cpp]: `#include` works as expected. If there are warnings or errors in an included file, the filename I print is the name of the top level frag or vert file. Line numbers refer to the preprocessed file after all `#include` statements have been expanded. If you want this fixed, it's a big deal and needs its own issue.

**Comment by doug-moen** [comment]
PR #54 is a simplified version of this PR in which I removed the code that you were concerned about.

**Comment by doug-moen** [comment]
I rebased this old PR on master. The only changes left implement the following change:

    If there are warnings or errors in one of the two possible shader files that you can
    specify on the command line (*.frag or *.vert), then the pathname of the shader
    file is printed ahead of the warnings or errors. If the shader being compiled is not
    one of these two possible shader files, then no pathname is printed.

    This is useful in the case where you have supplied both kinds of shader on the
    command line, and you want to know which of these two files has warnings or
    errors.

In order to implement this, I need to pass the pathnames of the shaders to shader::load and shader::compileShader. A _path argument is null if that shader is not one of the two shader files that can be specified on the command line.

Within compileShader(), I also test if _path is null to optimize some of the extra processing that is done for user-supplied shaders, such as automatic detection of a shadertoy.com shader.

This is a large number of changes to make for a relatively small benefit, so I will understand if you don't want this change in glslShader. I don't really need this change myself, since I don't currently plan to use vertex shaders for my project.

Just close this PR if you don't want the change.

**Comment by doug-moen** [comment]
Update, I found a simpler way to do this.

In Shader::compileShader(), I remove all of the error handling. The end of the function becomes:
```
    glCompileShader(shader);
    return shader;
}
```

In Shader::load(), I call glGetProgramInfoLog(m_program, ...) regardless of whether there is a link error or not. If there is a message, I print it.

The result is, the warnings and errors for the fragment and vertex shader are separated out and labeled by OpenGL. It looks like this for nVidia, with a vertex error and a fragment warning:
```
Errors while loading shaders:
Vertex info
-----------
(0) : error C0000: syntax error, unexpected $end, expecting "::" at token "<EOF>"
(0) : error C2003: incompatible options for link

Fragment info
-------------
0(72) : warning C7011: implicit cast from "int" to "float"

0: NOT FOUND
```

This is a much simpler change than the current PR, and it also prints link warnings if they exist.

**Comment by patriciogonzalezvivo** [comment]
this sound great! Thanks again! Still don't know your twitter and website... for adding it to the https://github.com/patriciogonzalezvivo/glslViewer/blob/master/README.md#acknowledgements Do you mind passing them or writing them there?
